### DISCUSSION TITLE
General purpose GPU accelerated parallel programming

### DISCUSSION THREAD

**Discussion by spamegg1** [bytedeco/storch#75] [Q&A] [answered]
I looked around but could not find an easy way to take advantage of the GPU in Scala. There is the [Parallel Collections library](https://github.com/scala/scala-parallel-collections/) but that runs on the CPU as far as I know.

Is it possible to use Storch for non-deep learning, non-numerical, more general parallel computation purposes? Like using the GPU to run a parallel blur filter that can run thousands of times faster than on the CPU, for example. This would be fantastic for educational purposes.

Are there any other "general parallel programming on the GPU" libraries for Scala that are easy to use and "teachable"?

**Comment by hmf** [comment]
I will defer this to @sbrunk, but to my knowledge I would say no because storch "just" acts as a wrapper to the JavaCCP wrapper of the Pytorch library. I am assuming here that one would need to program the kernels directly for such an exercise. A quick search also shows image blur is only available in [PytorchVision](https://pytorch.org/vision/main/generated/torchvision.transforms.GaussianBlur.html). I also did not find any [JavaCPP preset](https://github.com/bytedeco/javacpp-presets) for torchvision, but I may have missed it.

**Reply by spamegg1** [reply]
Thanks! 

I found [this](https://github.com/Syncleus/aparapi) but it hasn't been updated in 3 years...

There is also [this](https://github.com/pityka/lamp) but like Storch it seems to be deep-learning sci-comp oriented.

Once I venture into Java land I can find stuff like [this](https://github.com/gpu/JOCL). Not sure how easy / hard it is to use it in Scala. Also OpenCL setup is a massive pain on Linux.

Any suggestions, or opinions about the above, are welcome...

**Reply by sbrunk** [reply]
Its true that PyTorch/LibTorch and therefore Storch is optimized for building neural networks by providing automatic differentiation, higher level NN primitives etc. But you could totally build a fast GPU enabled blur filter running on the GPU using Storch tensor primitives.

As @hmf said, we don't have higher level operations like in TorchVision because they are mostly written in Python.

It could be a fun exercise to port i.e. a blur filter from there or some NumPy/SciPy implementation. Happy to help if you want to give it a try.

But that's still numerical right? For more general-purpose and especially non-numerical parallel computing it's not the best fit I guess. PyTorch tensors are always numeric.

Perhaps have a look at https://www.tornadovm.org/ although I don't know how well it works with Scala.

**Reply by spamegg1** [reply]
Thanks a lot! I might take you up on the offer, but maybe later :smile: 

One goal I'm interested in is to squeeze the most out of affordable APUs, like my Ryzen 5 3400G with 4c/8t CPU and 11c GPU. There are also many new laptops with nice AMD APUs these days. So students can get a decent amount of performance (like 10-15x parallel) out of a reasonably affordable single PC.

**Reply by hmf** [reply]
Small note on AMD's hardware - many frameworks (such as PyTorch and Tensorflow) depend on NVIDIA's CUDA library. We now have [ZLUDA](https://github.com/vosen/ZLUDA), although it is still very much a work in progress. Having said that, maybe it is not the best route to take.

**Reply by spamegg1** [reply]
Yeah, the whole Nvidia monopoly on GPU programming really sucks... it's hard to require someone to buy Nvidia graphics card just to program. I'll have to look at OpenCL options probably.

**Reply by hmf** [reply]
Don't forget TornadoVM - splash page says they support AMD.  8-) I had looked at this some time ago. Will take a look at [it](https://www.youtube.com/watch?v=VTzGlnv6nuA) again.

**Reply by spamegg1** [reply]
Yeah looks promising and supports OpenCL. The first example in the video is a blur filter too! :+1: 

But it might not run on AMD integrated GPUs (which is what I have):

> TornadoVM can currently execute on multi-core CPUs, dedicated GPUs (Nvidia, AMD), integrated GPUs (Intel HD Graphics and ARM Mali), and FPGAs (Intel and Xilinx).

**Reply by hmf** [reply]
Bummer. Maybe a student project to add that to the TornadoVM? ;-)

**Reply by spamegg1** [reply]
Immediately [hit a bug!](https://github.com/beehive-lab/TornadoVM/issues/391) :rofl: It is very much a work in progress indeed.

I think it won't work anyway; it requires OpenCL >= 2.1 which needs different hardware than mine (1.1).

Well, at least it built successfully. I'll try some Java code. Then figure out how to do it in Scala!

**Reply by spamegg1** [reply]
Maybe it will work? :thinking: 

```bash
╰─ tornado --devices                                                                                                 ─╯
WARNING: Using incubator modules: jdk.incubator.vector

Number of Tornado drivers: 1
Driver: OpenCL
  Total number of OpenCL devices  : 2
  Tornado device=0:0  (DEFAULT)
	OPENCL --  [Clover] -- AMD Radeon Vega 11 Graphics (raven, LLVM 15.0.7, DRM 3.54, 6.5.0-28-generic)
		Global Memory Size: 6.8 GB
		Local Memory Size: 64.0 KB
		Workgroup Dimensions: 3
		Total Number of Block Threads: [256]
		Max WorkGroup Configuration: [256, 256, 256]
		Device OpenCL C version: OpenCL C 1.1

  Tornado device=0:1
	OPENCL --  [Portable Computing Language] -- pthread-AMD Ryzen 5 3400G with Radeon Vega Graphics
		Global Memory Size: 11.6 GB
		Local Memory Size: 512.0 KB
		Workgroup Dimensions: 3
		Total Number of Block Threads: [4096]
		Max WorkGroup Configuration: [4096, 4096, 4096]
		Device OpenCL C version: OpenCL C 1.2 pocl
```

**Reply by sbrunk** [reply]
PyTorch also supports AMD GPUs via rocm BTW: https://pytorch.org/blog/pytorch-for-amd-rocm-platform-now-available-as-python-package/

I don't know how well APUs are supported though. Might be worth a shot trying to install the upstream PyTorch rocm variant via pip and test if it's working on your hardware.

**Reply by spamegg1** [reply, marked as answer]
Tornado doesn't work unfortunately :cry: 
Running the examples just stalls with 0% CPU usage.
Needs newer hardware I think! 
(Although the guy in the video runs it on a laptop. It's Nvidia... :angry: )

@sbrunk rocm is only supported on Instinct GPUs.

I think I give up on this dream :cry: 
GPU compute is reserved for the rich, no low-end peasants allowed.

**Reply by hmf** [reply]
Wouldn't hurt to ask in the [TVM slack channel](https://github.com/beehive-lab/TornadoVM/discussions) if the APU is supported. Quick search didn't not show anything. 

It just occurred to me that TVM may also be a way for storch to use GPU for example when loading and augmenting images.

**Comment by spamegg1** [comment]
I found this! https://github.com/ComputeNode/scalag

Did any of you know about this? Seems promising! 

Says SPIR-V and run through Vulkan, not OpenCL or CUDA. That can work on APUs, hopefully!

It's pre-alpha so I'll wait.

**Comment by hmf** [comment]
Not me. Seems interesting. And that image does not seem to be just ray-tracing. Sift shadows reminds me of the [Radiosity](https://en.wikipedia.org/wiki/Radiosity_(computer_graphics).
